i_1122 Create a catalog file during execution

johnbrvc opened the pull request:
### Description of what the PR does
The `src_catalog.json` file contains details about the contestant's source code, such as the original full path of the submission, whether the file is stored externally, etc. This is useful for generating specific web pages for the judge's submissions.  For example, one can look at the full path of the submission source and see if it's from the CDP (for example); if so, then we know they're judge's sample submissions.  

### Issue which the PR addresses
Fixes #1122 

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Windows 11
java version "1.8.0_321"
Java(TM) SE Runtime Environment (build 1.8.0_321-b07)
Java HotSpot(TM) 64-Bit Server VM (build 25.321-b07, mixed mode)

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)

1. Load a contest that has sample submissions from the judges in the CDP (ex. **NAC2025**)
2. Make sure the `pc2v9.ini` file has the `[admin].sampleSubmitPane` key set to true.  This will make the **Submit Samples** tab show up on the administrator's **Run Contest** tab.
3. Start the server.
4. Start an administrator.
5. Start the autojudges.
6. Start the contest.
7. Using the **Submit Samples** tab, filter by a single problem or two (so you don't have to wait for the system to judge all the judge's submissions.
8. Submit the judge's submissions and wait for them to be judged.
9. Look in the execute folder for one or more of the judgments (The execute folder will be named something like: `ex_6_A_atotientquotient_1_cpp_judge1`) for a file with the name `src_catalog.json`
10. It should contain detailed information about source file(s) for the submission.

johnbrvc committed:
i_1122 Create a catalog file during execution

The src_catalog.json file contains details about the contestant's source code, such as the original full path of the submission, whether the file is stored externally, etc.
This is useful for generating specific pages for the judge's submissions.  We can look at the full path of the submission source and see if it's from the CDP; if so, then we know they're judge's sample submissions.

johnbrvc committed:
i_1122 Fix comments and log errors

Pass a Log object to the catalog writing method in case of errors.

clevengr commented on src/edu/csus/ecs/pc2/core/execute/Executable.java:
```diff
@@ -2760,6 +2767,12 @@ protected boolean compileProgram() {
+            if(runFiles != null) {
+                runFiles.createCatalogJSON(catlog, log);
+                catlog.flush();
+            }
```
Shouldn't there be a `catlog.close()` here?

JoeTerlizzi replied:
Use try-with-resources to ensure close is called even if an exception occurs.

The try-with-resources statement in Java is a language construct introduced in Java 7 to simplify resource management, especially for resources that need to be closed after use (e.g., file streams, database connections, network sockets). It ensures that each resource declared within the try statement's parentheses is closed automatically when the try block exits, regardless of whether an exception occurs.

johnbrvc replied:
Changed `catlog.flush()` to `catlog.close()`.
I don't think we need the extravagance of the _try-with-resources_.  If it fails to create the file, the we don't have to clean anything up, and it will exception itself out up the chain (which is fine).  If the write fails (`createCatalogJSON()`), that catches it and deals with it gracefully.

clevengr commented on src/edu/csus/ecs/pc2/core/model/RunFiles.java:
```diff
@@ -178,16 +182,88 @@ public String toString() {
+                catlog.newLine();
+                catlog.write(" {\"mainfile\":");
+                catlog.write(serializedFileToJSON(mainFile));
+                catlog.write(" }");
```
I don't see the need for a space in front of the `}` in the above line...

johnbrvc replied:
I just had it there for readability of the json by a human.  I can live without it.  Removed.

clevengr commented on src/edu/csus/ecs/pc2/core/model/RunFiles.java:
```diff
@@ -178,16 +182,88 @@ public String toString() {
+                needComma = true;
+            }
+            if(otherFiles != null && otherFiles.length > 0) {
+                boolean needComma2 = false;
```
I don't see any reference/use of `needComma2` in the code.  Am I missing something?

JoeTerlizzi replied:
Agreed.  Remove this line.

johnbrvc replied:
This was a bug in my loop.  The loop should have used `needcomma2 `not `needcomma`.   Basically, if there's more than one "other file", then we have to put commas between those.  The other files are in a separate json object, so they have their own separators (`needComma2`).

I fixed the loop.

clevengr commented on src/edu/csus/ecs/pc2/core/model/RunFiles.java:
```diff
@@ -178,16 +182,88 @@ public String toString() {
+     * @param f the serialized file
+     * @return JSON string representation of the detailed info
+     */
+    private String serializedFileToJSON(SerializedFile f) {
```
The name of this method seems a bit misleading; it implies that it's creating a JSON string for the serialized file (whereas what it's actually doing is creating a JSON string for the "detailed info" of the serialized file).

johnbrvc replied:
Changed to `serializedFileInfoToJSON()`.

clevengr reviewed (commented):
I reviewed the code and left minor comments; nothing critical.  I pulled the code from the author's fork, built the system locally, and ran the tests described in the PR.  I see what appears to be correct `src_catalog.json` files in the selected (filtered) problems.  I also see submissions for exactly the submission files contained in the selected problems' `submissions` folders.

However, I do not see "expected results" in the "Status" field on the Submit Samples pane, and I likewise don't see expected judgements on the Runs screen.  Rather, except for the couple that got CE (which are explainable due to the different platform), all the submissions ended up in the "NEW" state.  Somethin is wrong somewhere, possibly (probably?) with my configuration.

clevengr reviewed (approved):
After struggling greatly with my local configuration, I finally got it correct and was able to properly execute the PR.  I get mismatches in the actual judge submission "Expected results" vs. "Status" -- compiler errors for one (which isn't surprising at all), plus several TLEs where "AC" was expected.  But again, those are explainable.  

And none of that is actually what the PR is about; it's about producing the proper `src_catalog.json` files in the execute directory.  I verified that this works, so I'm approving the PR.

JoeTerlizzi reviewed (approved):
I ran with NAC2025 contest; Resubmitted solutions to be autojudged; src_catalog.json had the correct abs path for source code.

I agree with @clevengr RE his two comments and left my own feedback there.  Assuming these are addressed...approved.

johnbrvc committed:
i_1122 PR Changes requests

Minor changes to address PR comments.
1) close cat file (shouldn't be necessary since when it goes out of scope it should close
2) Rename serializedFileToJSON() method to serializedFileInfoToJSON()
3) Fix loop that adds Other Files to source submissions in catalog.   Was not handling commas properly for separating multiple files.
